File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return
        
        prev = None
        curr = head
        fast = head.next
        
        while fast:
            curr.next = prev
            prev = curr
            curr = fast
            fast = fast.next
            
        curr.next = prev
        
        return curr
    
#3 pointers
# TC :O(n)
# SC: O(1). No extra space

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        Stack<ListNode> st = new Stack<ListNode>();

        ListNode curr = head.next;
        
        while(curr != null){
            head.next = null;
            st.push(head);
            head = curr;
            curr = curr.next;
        }

        curr = head;

        while(!st.isEmpty()){
            curr.next = st.pop();
            curr = curr.next;
        }

        return head;        
    }
}
// Using Stack
// Time Complexity: O(n)
// Space Complexity: O(n)